[TRRFDA-301] Update Demographic fields from Admin interface

alintreznai opened the pull request:
This PR allows an admin to show/hide demographic fields using the admin interface.

Commit:
Update DemographicsFields model, generate migrations

Commit:
Update demographic fields admin form

Commit:
Handle hidden sections in patient view

Commit:
Updated import and export for DemographicsFields

Commit:
Small cleanup

Commit:
Handle hidden fields in patient view

Commit:
Add is_section field in DemographicsFields export/import

Commit:
Code cleanup

sztamas commented on rdrf/rdrf/models/definition/models.py:
```diff
@@ -1389,12 +1389,14 @@ class ConsentRule(models.Model):
 
     registry = models.ForeignKey(Registry, on_delete=models.CASCADE)
-    group = models.ForeignKey(Group, on_delete=models.CASCADE)
+    groups = models.ManyToManyField(Group, related_name='groups')
```
`related_name` should be `demographic_fields`

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/services/io/defs/importer.py:
```diff
@@ -834,16 +834,25 @@ def _create_demographic_fields(self, data):
-                logger.info("created Group %s" % group_obj)
-                group_obj.save()
-
+            groups = []
```
Just to make it more visible on what's different between the 2 cases:

```python
group_names = d.get('groups', []])
# Support for importing data in the old format. We used to have demographic fields assigned to only one group in the past.
if 'group' in d:
    group_names = [d['group']]
for group_name in group_names:
    ...
```

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/views/patient_view.py:
```diff
@@ -25,6 +25,11 @@
 from django.utils.translation import ugettext as _
 
 from rdrf.forms.dynamic.registry_specific_fields import RegistrySpecificFieldsHandler
+from rdrf.helpers.constants import (
```
All these look like very patient form specific constants.
I would prefer keeping them local in this file instead of having them in a more general location like `rdrf.helpers.constants` and then importing them.
Any reason not to do so?

alintreznai replied:
I'm also using them in rdrf/admin_forms so I wanted to put them in a common location.

sztamas commented on rdrf/rdrf/views/patient_view.py:
```diff
@@ -539,22 +547,44 @@ def get_forms(self, request, registry_model, user, instance=None):
 
-    def _section_hidden(self, user, registry, fieldlist):
+    def _exclude_hidden_fields(self, user, registry, fieldlist):
+        from rdrf.models.definition.models import DemographicFields
```
Can we please move imports to the top of the file as a general rule?

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/views/patient_view.py:
```diff
@@ -539,22 +547,44 @@ def get_forms(self, request, registry_model, user, instance=None):
-    def _section_hidden(self, user, registry, fieldlist):
+    def _exclude_hidden_fields(self, user, registry, fieldlist):
+        from rdrf.models.definition.models import DemographicFields
+        user_groups = [g.name for g in user.groups.all()]
```
I don't think you have to get the group names. Just filter by `groups__in=user.groups.all()` instead.
Also applies to the other 2 "hidden" methods.

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/views/patient_view.py:
```diff
@@ -539,22 +547,44 @@ def get_forms(self, request, registry_model, user, instance=None):
+        ).values_list('field', flat=True)
+        return list(set(fieldlist) - set(hidden_fields))
+
+    def _section_fields_hidden(self, user, registry, fieldlist):
```
Shouldn't this be called `_all_section_fields_hidden`?
And why isn't the implementation just `len(self._exclude_hidden_fields(...)) == 0` ?

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/views/patient_view.py:
```diff
@@ -539,22 +547,44 @@ def get_forms(self, request, registry_model, user, instance=None):
+        ).exists()
+
     def _check_for_hidden_section(self, user, registry, form_sections):
         section_hiddenlist = []
```
All this nesting `initialisation`, `looping`, `continue`, `appending` isn't just a `filter` operation in disguise? :-)

alintreznai replied:
Updated.

sztamas commented on rdrf/rdrf/models/definition/models.py:
```diff
@@ -1389,12 +1389,14 @@ class ConsentRule(models.Model):
 
 class DemographicFields(models.Model):
     FIELD_CHOICES = []
+    SECTION_PREFIX = "SECTION:"
```
Wouldn't this be better placed on the form instead? Looks more like form than model functionality to me...

alintreznai replied:
Moved to the admin form.

Commit:
Updates based on code review

sztamas commented:
django.core.exceptions.FieldError
django.core.exceptions.FieldError: Cannot resolve keyword 'group' into field. Choices are: field, groups, hidden, id, is_section, readonly, registry, registry_id

at: 

File "/app/rdrf/registry/patients/admin_forms.py", line 243, in __init__

Commit:
Small fix

alintreznai commented:
Fixed the issue.

Commit:
Maintain order of fields when excluding hidden fields

Commit:
Updated testing zips

sztamas committed:
Fixes issue causing the failure of end2end tests.

+ some other small improvements.

Got rid of the exclude_hidden_fields methods as the fields are hidden by
including them, but setting their widgets to hidden instead.

sztamas commented:
@alintreznai fixed the issue that was causing the end2end test failures + made some other small improvements.
Please take a look.
Other things on my TODO list I didn't get to:

- shouldn't we make the `field` field on DemographicFields unique for each registry?
- prevent read-only being set on a Section. We could support it at a later stage, but currently not allowing it should be enough
- validate hidden and read-only on a field. If hidden is set don't allow read-only
- maybe as a separate small feature (to avoid having to mess more with migrations) change hidden and read-only on DemographicField to be BooleanField with default of False to stay consistent (same as is_section). Or maybe change it to a choice field instead. Then you could have either hidden or read-only but not both. You could have properties on the model to keep the code the same

Commit:
Implement features on the todo list

Commit:
Updated testing zips

Commit:
Required patient fields can't be made hidden or read-only

Commit:
Change allows_null to allows_blank in DemographicFieldsAdminForm

sztamas committed:
Small changes
